File: python_interface/tests/coeff_test.c
/*
 *
 *  This source file is part of ELINA (ETH LIbrary for Numerical Analysis).
 *  ELINA is Copyright © 2018 Department of Computer Science, ETH Zurich
 *  This software is distributed under GNU Lesser General Public License
 * Version 3.0. For more information, see the ELINA project website at:
 *  http://elina.ethz.ch
 *
 *  THE SOFTWARE IS PROVIDED "AS-IS" WITHOUT ANY WARRANTY OF ANY KIND, EITHER
 *  EXPRESS, IMPLIED OR STATUTORY, INCLUDING BUT NOT LIMITED TO ANY WARRANTY
 *  THAT THE SOFTWARE WILL CONFORM TO SPECIFICATIONS OR BE ERROR-FREE AND ANY
 *  IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE,
 *  TITLE, OR NON-INFRINGEMENT.  IN NO EVENT SHALL ETH ZURICH BE LIABLE FOR ANY
 *  DAMAGES, INCLUDING BUT NOT LIMITED TO DIRECT, INDIRECT,
 *  SPECIAL OR CONSEQUENTIAL DAMAGES, ARISING OUT OF, RESULTING FROM, OR IN
 *  ANY WAY CONNECTED WITH THIS SOFTWARE (WHETHER OR NOT BASED UPON WARRANTY,
 *  CONTRACT, TORT OR OTHERWISE).
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include "elina_coeff.h"


void test_set_scalar_int(elina_coeff_t *coeff){
	long int num = rand()%100;
	elina_coeff_set_scalar_int(coeff,num);
	printf("set scalar int : %ld coeff: ",num);
	elina_coeff_print(coeff);
	printf(" is zero: %d\n", elina_coeff_zero(coeff));
}


void test_set_scalar_mpq(elina_coeff_t *coeff){
	mpq_t mpq;
	mpq_init(mpq);
	long int p = rand()%1000;
	unsigned long int q= rand()%20+1;
	mpq_set_si(mpq,p,q);
	elina_coeff_set_scalar_mpq(coeff,mpq);
	printf("set scalar mpq : ");
	mpq_out_str(stdout,10,mpq);
	printf(" coeff: ");
	elina_coeff_print(coeff);
	printf(" is zero: %d\n", elina_coeff_zero(coeff));
	mpq_clear(mpq);
}

void test_set_scalar_frac(elina_coeff_t *coeff){
	long int p = rand()%100;
	unsigned long int q = rand()%20+1;
	elina_coeff_set_scalar_frac(coeff,p,q);
	printf("set scalar frac p : %ld q: %ld coeff: ",p,q);
	elina_coeff_print(coeff);
	printf("\n");
}


void test_set_scalar_double(elina_coeff_t *coeff){
	double d = (double)rand()/RAND_MAX*2.0-1.0;
	elina_coeff_set_scalar_double(coeff,d);
	printf("set scalar double : %g coeff: ",d);
	elina_coeff_print(coeff);
	printf(" is zero: %d\n", elina_coeff_zero(coeff));
}


void test_set_scalar_mpfr(elina_coeff_t *coeff){
	mpfr_t mpfr;
	mpfr_init(mpfr);
	double d = (double)rand()/RAND_MAX*2.0-1.0;
	mpfr_set_d(mpfr,d,GMP_RNDU);
	elina_coeff_set_scalar_mpfr(coeff,mpfr);
	printf("set scalar mpfr: ");
	mpfr_out_str(stdout,10,elina_scalar_print_prec,mpfr,GMP_RNDU);
	printf(" coeff: ");
	elina_coeff_print(coeff);
	printf(" is zero: %d\n", elina_coeff_zero(coeff));
	mpfr_clear(mpfr);
}


void test_set_interval_int(elina_coeff_t *coeff){
	long int inf = rand()%100;
	long int sup = rand()%100;
	elina_coeff_set_interval_int(coeff,inf,sup);
	printf("set interval int inf: %ld sup: %ld coeff: ",inf,sup);
	elina_coeff_print(coeff);
	printf(" is zero: %d\n", elina_coeff_zero(coeff));
}


void test_set_interval_mpq(elina_coeff_t *coeff){
	mpq_t inf1,sup1;
	mpq_init(inf1);
	mpq_init(sup1);
	long int p = rand()%1000;
	unsigned long int q= rand()%20+1;
	mpq_set_si(inf1,p,q);
	p = p + rand()%1000;
	mpq_set_si(sup1,p,q);
	elina_coeff_set_interval_mpq(coeff,inf1,sup1);
	printf("set interval mpq inf: ");
	mpq_out_str(stdout,10,inf1);
	printf(" sup: ");
	mpq_out_str(stdout,10,sup1);
	printf(" coeff: ");
	elina_coeff_print(coeff);
	printf(" is zero: %d\n", elina_coeff_zero(coeff));
	mpq_clear(inf1);
	mpq_clear(sup1);
}

void test_set_interval_frac(elina_coeff_t * coeff){
	long int p1 = rand()%100;
	unsigned long int q1 = rand()%20+1;
	long int p2 = rand()%100;
	unsigned long int q2 = rand()%20+1;
	elina_coeff_set_interval_frac(coeff,p1,q1,p2,q2);
	printf("set interval frac p1: %ld q1: %ld p2: %ld q2: %ld coeff: ",p1,q1,p2,q2);
	elina_coeff_print(coeff);
	printf(" is zero: %d\n", elina_coeff_zero(coeff));
}


void test_set_interval_double(elina_coeff_t *coeff){
	double inf = (double)rand()/RAND_MAX*2.0-1.0;
	double sup = (double)rand()/RAND_MAX*2.0-1.0;
	elina_coeff_set_interval_double(coeff,inf,sup);
	printf("set interval double inf: %g sup: %g coeff: ",inf,sup);
	elina_coeff_print(coeff);
	printf(" is zero: %d\n", elina_coeff_zero(coeff));
}


void test_set_interval_mpfr(elina_coeff_t * coeff){
	mpfr_t inf,sup;
	mpfr_init(inf);
	mpfr_init(sup);
	double d = (double)rand()/RAND_MAX*2.0-1.0;
	mpfr_set_d(inf,d,GMP_RNDU);
	
	d = (double)rand()/RAND_MAX*2.0-1.0;
	mpfr_set_d(sup,d,GMP_RNDU);

	elina_coeff_set_interval_mpfr(coeff,inf,sup);
	printf("set interval mpfr: ");
	mpfr_out_str(stdout,10,elina_scalar_print_prec,inf,GMP_RNDU);
	printf(" ");
	mpfr_out_str(stdout,10,elina_scalar_print_prec,sup,GMP_RNDU);
	printf(" coeff: ");
	elina_coeff_print(coeff);
	printf(" is zero: %d\n", elina_coeff_zero(coeff));
	mpfr_clear(inf);
	mpfr_clear(sup);
}	



void test_coeff_cmp(elina_coeff_t * coeff1, elina_coeff_t *coeff2){
	elina_coeff_t *coeff3 = elina_coeff_alloc(ELINA_COEFF_SCALAR);
	elina_coeff_t *coeff4 = elina_coeff_alloc(ELINA_COEFF_INTERVAL);
	int c = rand()%5;
	switch(c){
		case 0:
			test_set_scalar_int(coeff3);
			break;
		case 1:
			test_set_scalar_mpq(coeff3);
			break;
		case 2:
			test_set_scalar_frac(coeff3);
			break;
		case 3:
			test_set_scalar_double(coeff3);
			break;
		default:
			test_set_scalar_mpfr(coeff3);
			break;
	}

	c = rand()%5;
	switch(c){
		case 0:
			test_set_interval_int(coeff4);
			break;
		case 1:
			test_set_interval_mpq(coeff4);
			break;
		case 2:
			test_set_interval_frac(coeff4);
			break;
		case 3:
			test_set_interval_double(coeff4);
			break;
		default:
			test_set_interval_mpfr(coeff4);
			break;
	}

	printf("cmp scalar vs interval coeff: ");
	elina_coeff_print(coeff1);
	printf(" ");
	elina_coeff_print(coeff2);
	printf(" coeff1<=coeff2: %d coeff2<=coeff1: %d\n",elina_coeff_cmp(coeff1,coeff2), elina_coeff_cmp(coeff2,coeff1));


	printf("cmp scalar coeff: ");
	elina_coeff_print(coeff1);
	printf(" ");
	elina_coeff_print(coeff3);
	printf(" coeff1<=coeff2: %d coeff2<=coeff1: %d\n",elina_coeff_cmp(coeff1,coeff3), elina_coeff_cmp(coeff3,coeff1));

	printf("cmp interval coeff: ");
	elina_coeff_print(coeff4);
	printf(" ");
	elina_coeff_print(coeff2);
	printf(" coeff1<=coeff2: %d coeff2<=coeff1: %d\n",elina_coeff_cmp(coeff4,coeff2), elina_coeff_cmp(coeff2,coeff4));

	elina_coeff_free(coeff3);
	elina_coeff_free(coeff4);
}


void test_coeff_equality(elina_coeff_t * coeff1, elina_coeff_t *coeff2){
	
	
	elina_coeff_t *coeff3 = elina_coeff_alloc(ELINA_COEFF_SCALAR);
	elina_coeff_t *coeff4 = elina_coeff_alloc(ELINA_COEFF_INTERVAL);
	int c = rand()%5;
	switch(c){
		case 0:
			test_set_scalar_int(coeff3);
			break;
		case 1:
			test_set_scalar_mpq(coeff3);
			break;
		case 2:
			test_set_scalar_frac(coeff3);
			break;
		case 3:
			test_set_scalar_double(coeff3);
			break;
		default:
			test_set_scalar_mpfr(coeff3);
			break;
	}

	c = rand()%5;
	switch(c){
		case 0:
			test_set_interval_int(coeff4);
			break;
		case 1:
			test_set_interval_mpq(coeff4);
			break;
		case 2:
			test_set_interval_frac(coeff4);
			break;
		case 3:
			test_set_interval_double(coeff4);
			break;
		default:
			test_set_interval_mpfr(coeff4);
			break;
	}

	printf("equal scalar vs interval coeff: ");
	elina_coeff_print(coeff1);
	printf(" ");
	elina_coeff_print(coeff2);
	printf(" coeff1==coeff2: %d\n",elina_coeff_equal(coeff1,coeff2));

	printf("equal scalar coeff: ");
	elina_coeff_print(coeff1);
	printf(" ");
	elina_coeff_print(coeff3);
	printf(" coeff1==coeff2: %d\n",elina_coeff_equal(coeff1,coeff3));

	printf("equal interval coeff: ");
	elina_coeff_print(coeff4);
	printf(" ");
	elina_coeff_print(coeff2);
	printf(" coeff1==coeff2: %d\n",elina_coeff_equal(coeff4,coeff2));

	elina_coeff_free(coeff3);
	elina_coeff_free(coeff4);
}

void test_coeff_neg(elina_coeff_t * coeff1, elina_coeff_t *coeff2){
	elina_coeff_t *coeff3 = elina_coeff_alloc(ELINA_COEFF_SCALAR);
	elina_coeff_t *coeff4 = elina_coeff_alloc(ELINA_COEFF_INTERVAL);
	elina_coeff_neg(coeff3,coeff1);
	printf("scalar coeff: ");
	elina_coeff_print(coeff1);
	printf("neg coeff: ");
	elina_coeff_print(coeff3);
	printf("\n");

	elina_coeff_neg(coeff4,coeff2);
	printf("interval coeff: ");
	elina_coeff_print(coeff2);
	printf("neg coeff: ");
	elina_coeff_print(coeff4);
	printf("\n");

	elina_coeff_free(coeff3);
	elina_coeff_free(coeff4);
}

void test_coeff_reduce(){
	elina_coeff_t * coeff = elina_coeff_alloc(ELINA_COEFF_INTERVAL);
	long int num = rand()%100;
	elina_coeff_set_interval_int(coeff,num,num);
	printf("Before reduce: ");
	elina_coeff_print(coeff);
	printf("\n");
	elina_coeff_reduce(coeff);
	printf("After reduce: ");
	elina_coeff_print(coeff);
	printf("\n");
	elina_coeff_free(coeff);
}





int main(){
	srand (time(NULL));

	elina_coeff_t * coeff1 = elina_coeff_alloc(ELINA_COEFF_SCALAR);

	elina_coeff_t * coeff2 = elina_coeff_alloc(ELINA_COEFF_INTERVAL);

	// coeff is set to scalar of type int
	test_set_scalar_int(coeff1);
	
	//coeff is set to scalar of type mpq
	test_set_scalar_mpq(coeff1);
	
	//coeff is set to scalar of type fraction of two integers
	test_set_scalar_frac(coeff1);
	
	//coeff is set to scalar of type double
	test_set_scalar_double(coeff1);
	
	//coeff is set to scalar of type mpfr
	test_set_scalar_mpfr(coeff1);	
	
	// coeff is set to interval of type int
	test_set_interval_int(coeff2);
	
	//coeff is set to interval of type mpq
	test_set_interval_mpq(coeff2);
	
	//coeff is set to interval of type fraction of two integers
	test_set_interval_frac(coeff2);
	
	//coeff is set to interval of type double
	test_set_interval_double(coeff2);
	
	//coeff is set to interval of type mpfr
	test_set_interval_mpfr(coeff2);

	//set to coeff
	//test_set_coeff(coeff1, coeff2);	
	
	//test for comparison and equality 
	test_coeff_cmp(coeff1, coeff2);
	test_coeff_equality(coeff1, coeff2);
	
	
	//test for negation and reduce
	test_coeff_neg(coeff1, coeff2);
	test_coeff_reduce();


	elina_coeff_free(coeff1);
	elina_coeff_free(coeff2);
}

File: python_interface/tests/coeff_test.py
#
#
#  This source file is part of ELINA (ETH LIbrary for Numerical Analysis).
#  ELINA is Copyright © 2019 Department of Computer Science, ETH Zurich
#  This software is distributed under GNU Lesser General Public License Version 3.0.
#  For more information, see the ELINA project website at:
#  http://elina.ethz.ch
#
#  THE SOFTWARE IS PROVIDED "AS-IS" WITHOUT ANY WARRANTY OF ANY KIND, EITHER
#  EXPRESS, IMPLIED OR STATUTORY, INCLUDING BUT NOT LIMITED TO ANY WARRANTY
#  THAT THE SOFTWARE WILL CONFORM TO SPECIFICATIONS OR BE ERROR-FREE AND ANY
#  IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE,
#  TITLE, OR NON-INFRINGEMENT.  IN NO EVENT SHALL ETH ZURICH BE LIABLE FOR ANY     
#  DAMAGES, INCLUDING BUT NOT LIMITED TO DIRECT, INDIRECT,
#  SPECIAL OR CONSEQUENTIAL DAMAGES, ARISING OUT OF, RESULTING FROM, OR IN
#  ANY WAY CONNECTED WITH THIS SOFTWARE (WHETHER OR NOT BASED UPON WARRANTY,
#  CONTRACT, TORT OR OTHERWISE).
#
#


import sys
sys.path.insert(0, '../')

from elina_auxiliary_imports import *
from elina_coeff import *
from test_imports import *


def test_set_scalar_int(coeff):
    num = c_long(random.randint(0, 99))
    elina_coeff_set_scalar_int(coeff, num)
    print_c('set scalar int : {} coeff: '.format(num))
    elina_coeff_fprint(cstdout, coeff)
    print_c(' is zero: {}\n'.format(elina_coeff_zero(coeff)))


def test_set_scalar_mpq(coeff):
    mpq_t = Mpq_t()
    libgmp.__gmpq_init(mpq_t)
    p = c_long(random.randint(0, 999))
    q = c_ulong(random.randint(1, 20))
    libgmp.__gmpq_set_si(mpq_t, p, q)
    elina_coeff_set_scalar_mpq(coeff, mpq_t)
    print_c('set scalar mpq: ')
    libgmp.__gmpq_out_str(cstdout, 10, mpq_t)
    print_c(' coeff: ')
    elina_coeff_fprint(cstdout, coeff)
    print_c(' is zero: {}\n'.format(elina_coeff_zero(coeff)))
    libgmp.__gmpq_clear(mpq_t)


def test_set_scalar_frac(coeff):
    p = c_long(random.randint(0, 99))
    q = c_ulong(random.randint(1,20))
    elina_coeff_set_scalar_frac(coeff, p, q)
    print_c('set scalar frac p: {} q: {} coeff '.format(p, q))
    elina_coeff_fprint(cstdout, coeff)
    print_c('\n')


def test_set_scalar_double(coeff):
    d = c_double(random.uniform(-1, 1))
    elina_coeff_set_scalar_double(coeff, d)
    print_c('set scalar double: {} coeff: '.format(d))
    elina_coeff_fprint(cstdout, coeff)
    print_c(' is zero: {}\n'.format(elina_coeff_zero(coeff)))


def test_set_scalar_mpfr(coeff):
    mpfr_t = Mpfr_t()
    libmpfr.mpfr_init(mpfr_t)
    d = c_double(random.uniform(-1, 1))
    libmpfr.mpfr_set_d(mpfr_t, d, MpfrRnd.MPFR_RNDU)
    elina_coeff_set_scalar_mpfr(coeff, mpfr_t)
    print_c('set scalar mpfr: ')
    libmpfr.__gmpfr_out_str(cstdout, 10, elina_scalar_print_prec, mpfr_t, MpfrRnd.MPFR_RNDU)
    print_c(' coeff: ')
    elina_coeff_fprint(cstdout, coeff)
    print_c(' is zero: {}\n'.format(elina_coeff_zero(coeff)))
    libmpfr.mpfr_clear(mpfr_t)


def test_set_interval_int(coeff):
    inf = c_long(random.randint(0, 99))
    sup = c_long(random.randint(0, 99))
    elina_coeff_set_interval_int(coeff, inf, sup)
    print_c('set interval int inf: {} sup: {} coeff: '.format(inf, sup))
    elina_coeff_fprint(cstdout, coeff)
    print_c(' is zero: {}\n'.format(elina_coeff_zero(coeff)))


def test_set_interval_mpq(coeff):
    inf1 = Mpq_t()
    sup1 = Mpq_t()
    libgmp.__gmpq_init(inf1)
    libgmp.__gmpq_init(sup1)
    p = c_long(random.randint(0, 999))
    q = c_ulong(random.randint(1, 20))
    libgmp.__gmpq_set_si(inf1, p, q)
    p = c_long(p.value + random.randint(0, 999))
    libgmp.__gmpq_set_si(sup1, p, q)
    elina_coeff_set_interval_mpq(coeff, inf1, sup1)
    print_c('set interval mpq inf: ')
    libgmp.__gmpq_out_str(cstdout, 10, inf1)
    print_c(' sup: ')
    libgmp.__gmpq_out_str(cstdout, 10, sup1)
    print_c(' coeff: ')
    elina_coeff_fprint(cstdout, coeff)
    print_c(' is zero: {}\n'.format(elina_coeff_zero(coeff)))
    libgmp.__gmpq_clear(inf1)
    libgmp.__gmpq_clear(sup1)


def test_set_interval_frac(coeff):
    p1 = c_long(random.randint(0, 99))
    q1 = c_ulong(random.randint(1, 20))
    p2 = c_long(random.randint(0, 99))
    q2 = c_ulong(random.randint(1, 20))
    elina_coeff_set_interval_frac(coeff, p1, q1, p2, q2)
    print_c('set interval frac p1: {} q1: {} p2: {} q2: {} coeff: '.format(p1, q1, p2, q2))
    elina_coeff_fprint(cstdout, coeff)
    print_c(' is zero: {}\n'.format(elina_coeff_zero(coeff)))


def test_set_interval_double(coeff):
    inf = c_double(random.uniform(-1, 1))
    sup = c_double(random.uniform(-1, 1))
    elina_coeff_set_interval_double(coeff, inf, sup)
    print_c('set interval double inf: {} sup: {} coeff: '.format(inf, sup))
    elina_coeff_fprint(cstdout, coeff)
    print_c(' is zero: {}\n'.format(elina_coeff_zero(coeff)))


def test_set_interval_mpfr(coeff):
    inf = Mpfr_t()
    sup = Mpfr_t()
    libmpfr.mpfr_init(inf)
    libmpfr.mpfr_init(sup)
    d = c_double(random.uniform(-1, 1))
    libmpfr.mpfr_set_d(inf, d, MpfrRnd.MPFR_RNDU)
    d = c_double(random.uniform(-1, 1))
    libmpfr.mpfr_set_d(sup, d, MpfrRnd.MPFR_RNDU)

    elina_coeff_set_interval_mpfr(coeff, inf, sup)
    print_c('set interval mpfr inf: ')
    libmpfr.__gmpfr_out_str(cstdout, 10, elina_scalar_print_prec, inf, MpfrRnd.MPFR_RNDU)
    print_c(' sup: ')
    libmpfr.__gmpfr_out_str(cstdout, 10, elina_scalar_print_prec, sup, MpfrRnd.MPFR_RNDU)
    print_c(' coeff: ')
    elina_coeff_fprint(cstdout, coeff)
    print_c(' is zero: {}\n'.format(elina_coeff_zero(coeff)))
    libmpfr.mpfr_clear(inf)
    libmpfr.mpfr_clear(sup)


def test_coeff_cmp(coeff1, coeff2):
    print_c('Test coeff compare:\n')
    coeff3 = elina_coeff_alloc(ElinaCoeffDiscr.ELINA_COEFF_SCALAR)
    coeff4 = elina_coeff_alloc(ElinaCoeffDiscr.ELINA_COEFF_INTERVAL)

    c = c_int(random.randint(0, 4))
    if c == 0:
        test_set_scalar_int(coeff3)
    elif c == 1:
        test_set_scalar_mpq(coeff3)
    elif c == 2:
        test_set_scalar_frac(coeff3)
    elif c == 3:
        test_set_scalar_double(coeff3)
    else:
        test_set_scalar_mpfr(coeff3)

    c = c_int(random.randint(0, 4))
    if c == 0:
        test_set_interval_int(coeff4)
    elif c == 1:
        test_set_interval_mpq(coeff4)
    elif c == 2:
        test_set_interval_frac(coeff4)
    elif c == 3:
        test_set_interval_double(coeff4)
    else:
        test_set_interval_mpfr(coeff4)

    print_c('cmp scalar vs interval coeff: ')
    elina_coeff_fprint(cstdout, coeff1)
    print_c(' ')
    elina_coeff_fprint(cstdout, coeff2)
    print_c(' coeff1 <= coeff2: {} coeff2 <= coeff1: {}\n'.format(elina_coeff_cmp(coeff1, coeff2),
                                                                  elina_coeff_cmp(coeff2, coeff1)))

    print_c('cmp scalar coeff: ')
    elina_coeff_fprint(cstdout, coeff1)
    print_c(' ')
    elina_coeff_fprint(cstdout, coeff3)
    print_c(' coeff1 <= coeff3: {} coeff3 <= coeff1: {}\n'.format(elina_coeff_cmp(coeff1, coeff3),
                                                                  elina_coeff_cmp(coeff3, coeff1)))

    print_c('cmp interval coeff: ')
    elina_coeff_fprint(cstdout, coeff2)
    print_c(' ')
    elina_coeff_fprint(cstdout, coeff4)
    print_c(' coeff2 <= coeff4: {} coeff4 <= coeff2: {}\n'.format(elina_coeff_cmp(coeff2, coeff4),
                                                                  elina_coeff_cmp(coeff4, coeff2)))


def test_coeff_equality(coeff1, coeff2):
    print_c('Test coeff equality:\n')

    coeff3 = elina_coeff_alloc(ElinaCoeffDiscr.ELINA_COEFF_SCALAR)
    coeff4 = elina_coeff_alloc(ElinaCoeffDiscr.ELINA_COEFF_INTERVAL)

    c = c_int(random.randint(0, 4))
    if c == 0:
        test_set_scalar_int(coeff3)
    elif c == 1:
        test_set_scalar_mpq(coeff3)
    elif c == 2:
        test_set_scalar_frac(coeff3)
    elif c == 3:
        test_set_scalar_double(coeff3)
    else:
        test_set_scalar_mpfr(coeff3)

    c = c_int(random.randint(0, 4))
    if c == 0:
        test_set_interval_int(coeff4)
    elif c == 1:
        test_set_interval_mpq(coeff4)
    elif c == 2:
        test_set_interval_frac(coeff4)
    elif c == 3:
        test_set_interval_double(coeff4)
    else:
        test_set_interval_mpfr(coeff4)

    print_c('equal scalar vs interval coeff: ')
    elina_coeff_fprint(cstdout, coeff1)
    print_c(' ')
    elina_coeff_fprint(cstdout, coeff2)
    print_c(' coeff1 == coeff2: {}\n'.format(elina_coeff_equal(coeff1, coeff2)))

    print_c('equal scalar coeff: ')
    elina_coeff_fprint(cstdout, coeff1)
    print_c(' ')
    elina_coeff_fprint(cstdout, coeff3)
    print_c(' coeff1 == coeff3: {}\n'.format(elina_coeff_equal(coeff1, coeff3)))

    print_c('equal interval coeff: ')
    elina_coeff_fprint(cstdout, coeff2)
    print_c(' ')
    elina_coeff_fprint(cstdout, coeff4)
    print_c(' coeff2 == coeff4: {}\n'.format(elina_coeff_equal(coeff2, coeff4)))


def test_coeff_neg(coeff1):
    print_c('Test coeff neg:\n')
    coeff3 = elina_coeff_alloc(ElinaCoeffDiscr.ELINA_COEFF_SCALAR)
    elina_coeff_neg(coeff3, coeff1)
    print_c('scalar coeff: ')
    elina_coeff_fprint(cstdout, coeff1)
    print_c(' neg coeff: ')
    elina_coeff_fprint(cstdout, coeff3)
    print_c('\n')


def test_coeff_reduce():
    coeff = elina_coeff_alloc(ElinaCoeffDiscr.ELINA_COEFF_INTERVAL)
    num = c_long(random.randint(0,99))
    elina_coeff_set_interval_int(coeff, num, num)
    print_c('Before reduce: ')
    elina_coeff_fprint(cstdout, coeff)
    print_c('\n')
    elina_coeff_reduce(coeff)
    print_c('After reduce: ')
    elina_coeff_fprint(cstdout, coeff)
    print_c('\n')


coeff1 = elina_coeff_alloc(ElinaCoeffDiscr.ELINA_COEFF_SCALAR)
coeff2 = elina_coeff_alloc(ElinaCoeffDiscr.ELINA_COEFF_INTERVAL)

test_set_scalar_int(coeff1)
test_set_scalar_mpq(coeff1)
test_set_scalar_frac(coeff1)
test_set_scalar_double(coeff1)
test_set_scalar_mpfr(coeff1)

test_set_interval_int(coeff2)
test_set_interval_mpq(coeff2)
test_set_interval_frac(coeff2)
test_set_interval_double(coeff2)
test_set_interval_mpfr(coeff2)

test_coeff_cmp(coeff1, coeff2)
test_coeff_equality(coeff1, coeff2)
test_coeff_neg(coeff1)
test_coeff_reduce()
